### PR TITLE
feat(web): optimization via lazy construction of OSK layers ⏩

### PR DESCRIPTION
**Pull request by jahorton**
Continues from #11263, addressing part of #11166.

One particularly easy way to load and display the OSK more quickly is to simply... not make the parts you don't need yet.  We can do just-in-time (JIT) construction of OSK layers when they're needed.

While the `default` layer is pretty much always needed, and some keyboards will sometimes start in one or two other layers... a large number of keyboards have extra layers that can wait.

As there's one more followup on this set of PRs...

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by jahorton** [comment]
Current build-failure cause:

```
15:26:42   renderer_core.ts(181,54): error TS2341: Property 'layers' is private and only accessible within class 'OSKLayerGroup'.
```

Right, some of the tools will want access to that.

File: web/src/engine/osk/src/keyboard-layout/oskLayerGroup.ts [hunk 1]
```diff
@@ -47,31 +51,56 @@ export default class OSKLayerGroup {
...
+  }
 
-      // Add layer to group
-      lDiv.appendChild(layerObj.element);
+  public getLayer(id: string) {
+    let layer = this._layers[id];
+    if(!layer) {
+      layer = this.buildLayer(id);
+      if(layer) {
+        this._layers[id] = layer;
```
Comment (ermshiperete) [inline web/src/engine/osk/src/keyboard-layout/oskLayerGroup.ts:79]: This line seems unnecessary since that's already done in `buildLayer`.

**Review by ermshiperete — APPROVED** [review]

**Comment by jahorton** [comment]
Build failure seen for "Test: Language Modeling Layer (Common)":

```

15:56:57 
  LMLayer using the trie model

15:56:57 
    Prediction

15:56:57 
    Regression tests

15:56:57 
      should use the default searchTermToKey()

15:56:58 
        expected +0 to be at least 1

15:56:58 
        AssertionError: expected +0 to be at least 1
            at file:///C:/BuildAgent/work/99b311828f4ee7c/keyman/common/predictive-text/unit_tests/headless/worker-trie-integration.js:85:16
```

The worker's "test mode" is on, so we shouldn't be exiting early... but perhaps recent changes from #11784 and its predecessors have had an unintended consequence here.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.60-alpha](https://keyman.com/downloads/releases/alpha/18.0.60)

File: web/src/engine/osk/src/visualKeyboard.ts [hunk 2]
```diff
@@ -1316,13 +1316,13 @@ export default class VisualKeyboard extends EventEmitter<EventMap> implements Ke
...
 
-    const layers = this.layerGroup.layers;
+    const layers = this.layerGroup.spec.layer;
     let oskHeight = 0;
 
     // In case the keyboard's layers have differing row counts, we check them all for the maximum needed oskHeight.
     for (const layerID in layers) {
       const layer = layers[layerID];
-      let nRows = layer.rows.length;
+      let nRows = layer.row.length;
```
Comment (mcdurdin) [inline web/src/engine/osk/src/visualKeyboard.ts:1325]: Why this change?

**Review by mcdurdin — COMMENTED** [review]
Do you have any stats on improvement in performance?

**Comment by jahorton** [comment]
> Do you have any stats on improvement in performance?

Not in isolation.  I did collect stats for the set of three it was a part of though, and that can be found on https://github.com/keymanapp/keyman/pull/11265.

File: web/src/engine/osk/src/visualKeyboard.ts (hunk 2 above)
Comment (jahorton) [reply web/src/engine/osk/src/visualKeyboard.ts:1325]: Took a bit to find it, but the reason is on line 1319.  It shifts from using the set of _pre-built_ OSK layers (which have a `.rows` property) to using the underlying touch-layout specification (which uses `.row`, not `.rows`) as the basis for our calculations here.

Comment (mcdurdin) [reply web/src/engine/osk/src/visualKeyboard.ts:1325]: Understood. A bit squidgy, but make sense!
